Add extension commands and west.yml

mbolivar opened the pull request:
West now supports a mechanism for extension commands. Use it to move the command implementations that are tightly coupled with boards and the zephyr build system back into the Zephyr repository.

~~As a hack, for now this means they won't have any test cases, because those are part of the west repository. These will need to be resurrected later once a suitable mechanism is found to move them to zephyr's CI.~~

Edit: test cases restored.

@nashif -- in case you are wondering about the run_ci.sh changes, the new tests run very fast. Less than a second.

Commit:
scripts: add west build, flash, and debug commands

West now supports a mechanism for extension commands. Use it to move
the command implementations that are tightly coupled with boards and
the zephyr build system back into the Zephyr repository.

This patch doesn't include test cases. Those will be moved over in a
subsequent patch.

Signed-off-by: Marti Bolivar <marti@foundries.io>

carlescufi commented on scripts/west-commands.yml:
```diff
@@ -0,0 +1,17 @@
+west-commands:
```
Would it be better to have:
`scripts/west/west-commands.yml`
and then:
`scripts/west/commands/*.py`
?

mbolivar replied:
I made a change from when we initially discussed this feature: the extension commands can import west modules, but we no longer make any requirements on the file system layout.

See this comment: https://github.com/zephyrproject-rtos/west/blob/master/src/west/commands/command.py#L213

The reason I chose this layout was because "flat is better than nested" in the zen of python and there is no need to nest.

OK for you?

carlescufi replied:
That would be my preference, yes:
```
scripts/west/west-commands.yml
scripts/west/commands/*.py
```
@tejlmand can you comment on your preference as well to get more input?

mbolivar replied:
I just want to get this done and don't want to argue.

But I think having a `scripts/west` is dangerous: it might suggest that's where west lives. People will need time to understand that it's in its own repository.

But if that is the consensus I will make this change.

tejlmand replied:
i'm really fine with both ways as long as the python scripts them-selves is in sub-folder.

carlescufi replied:
Ok, I buy the arguments. Let's leave this as-is

tejlmand reviewed (approved):
Looks good to me.
Can see it's just existing code moved as-is, except for some adjusted import statements on debug.py and flash.py.

I assume that CI will become green on this, when reworked `west` is integrated into zephyr.

Commit:
scripts: run_ci.sh: remove dead code

Commit 88ece494 ("ci: do not build docs in main CI job") removed the
call to build_docs but not the function itself. That makes it
confusing to try to find where it's called; remove it.

Signed-off-by: Marti Bolivar <marti@foundries.io>

Commit:
scripts: run_ci.sh: add pytest-based testing for west commands

Incorporate these into run_ci.sh the same way that btsim results were
done. This adds a dependency on pytest.

Signed-off-by: Marti Bolivar <marti@foundries.io>

dbkinder commented on doc/tools/west/west-apis.rst:
```diff
@@ -5,6 +5,4 @@
 West APIs
 #########
 
-.. automodule:: west.runners.core
```
No documentation in this file now?

mbolivar replied:
Fixed now; had to do some spelunking into the latest CI system

Commit:
west: add west.yml

This is the west manifest file for mainline zephyr.

For a burn-in period, we are going to keep west pointing at master to
avoid deps rolls as we iron out any kinks.  We can then set its
revision to a SHA after things settle down.

The projects key is mandatory (if there's only one repository, there
wouldn't be much need for west!) so we'll start things off with the
net-tools repository at its current master SHA.

Signed-off-by: Marti Bolivar <marti@foundries.io>

tejlmand reviewed (changes requested):
Requesting to have `path = zephyr` present in `west.yml`

carlescufi commented on west.yml:
```diff
@@ -0,0 +1,23 @@
+      url-base: https://github.com/zephyrproject-rtos
+
+  self:
+    west-commands: scripts/west-commands.yml
```
+1 on this. Let's default to path `zephyr/` on the upstream manifest. Otherwise we'd need to change the doc substantially as well in case users choose a different repository name.

mbolivar replied:
Done

Commit:
run_ci.sh: HACK: install latest west

After some discussion, we've decided to ensure that shippable runs
with the latest west bootstrapper on PyPI by making that happen in
run_ci.sh, rather than updating our CI container. There are apparently
some issues going on with building that container, so this is a
shortest path option to making things green.

Signed-off-by: Marti Bolivar <marti@foundries.io>
